src/services/identity/UsersService.ts

```diff
@@ -230,56 +233,201 @@ class UsersService {
+  private verifyOtp(
+    otpEntry: Otp | null,
+    otp: string,
+    transaction: Transaction
+  ) {
     // TODO: Change all the following to use AuthError after FE fix
-    if (!otp || otp === "") {
-      throw new BadRequestError("Empty OTP provided")
-    }
-
-    if (!otpEntry) {
-      throw new BadRequestError("OTP not found")
-    }
-
-    if (otpEntry.attempts >= MAX_NUM_OTP_ATTEMPTS) {
-      throw new BadRequestError("Max number of attempts reached")
-    }
-
-    if (!otpEntry?.hashedOtp) {
-      await otpEntry.destroy()
-      throw new BadRequestError("Hashed OTP not found")
-    }
+    return okAsync(otp)
+      .andThen((otp) => {
+        if (!otp || otp === "") {
+          return errAsync(new BadRequestError("Empty OTP provided"))
+        }
 
-    // increment attempts
-    await otpEntry.update({ attempts: otpEntry.attempts + 1 })
-
-    const isValidOtp = await this.otpService.verifyOtp(otp, otpEntry.hashedOtp)
-    if (!isValidOtp) {
-      throw new BadRequestError("OTP is not valid")
-    }
+        return okAsync(otp)
+      })
+      .andThen(() => {
+        if (!otpEntry) {
+          return errAsync(new BadRequestError("OTP not found"))
+        }
 
-    if (isValidOtp && otpEntry.expiresAt < new Date()) {
-      await otpEntry.destroy()
-      throw new BadRequestError("OTP has expired")
-    }
+        return okAsync(otpEntry)
+      })
+      .andThen((otpDbEntry) => {
+        if (otpDbEntry.attempts >= MAX_NUM_OTP_ATTEMPTS) {
+          return errAsync(new BadRequestError("Max number of attempts reached"))
+        }
 
-    // destroy otp before returning true since otp has been "used"
-    await otpEntry.destroy()
-    return true
+        return okAsync(otpDbEntry)
+      })
+      .andThen((otpDbEntry) => {
+        if (!otpDbEntry.hashedOtp) {
+          return ResultAsync.fromPromise(
+            otpDbEntry.destroy({ transaction }),
+            (error) => {
+              logger.error(
+                `Error destroying OTP entry: ${JSON.stringify(error)}`
+              )
+
+              return new DatabaseError("Error destroying OTP entry in database")
+            }
+          ).andThen(() => errAsync(new BadRequestError("Hashed OTP not found")))
+        }
+
+        return okAsync(otpDbEntry)
+      })
+      .andThen((otpDbEntry) =>
+        // increment attempts
+        ResultAsync.fromPromise(
+          this.otpRepository.increment("attempts", {
+            where: { id: otpDbEntry.id },
+            transaction,
+          }),
+          (error) => {
+            logger.error(
+              `Error incrementing OTP attempts: ${JSON.stringify(error)}`
+            )
+
+            return new DatabaseError("Error incrementing OTP attempts")
+          }
+        ).map(() => otpDbEntry)
+      )
+      .andThen((otpDbEntry) =>
+        ResultAsync.combine([
+          okAsync(otpDbEntry),
+          ResultAsync.fromPromise(
+            this.otpService.verifyOtp(otp, otpDbEntry.hashedOtp),
```

Comment:
The verification on the number of attempts is only done here, and we increment beforehand because the number of attempts starts from 0. This means that the first attempt is when attempts = 1, then the fifth attempt is when attempts = 5. Thereafter, they can no longer attempt (so we limit users to 5 OTP attempts).